File: swift/Sources/FluentCards/MediaBuilder.swift
/// MediaBuilder builds a Media Adaptive Card element.
public final class MediaBuilder {
    var data: [String: Any] = ["type": "Media", "sources": [Any]()]

    public init() {}

    @discardableResult
    public func withID(_ id: String) -> MediaBuilder {
        data["id"] = id
        return self
    }

    @discardableResult
    public func withPoster(_ poster: String) -> MediaBuilder {
        data["poster"] = poster
        return self
    }

    @discardableResult
    public func withAltText(_ altText: String) -> MediaBuilder {
        data["altText"] = altText
        return self
    }

    @discardableResult
    public func withSpacing(_ spacing: Spacing) -> MediaBuilder {
        data["spacing"] = spacing.rawValue
        return self
    }

    /// Adds a media source with the given URL and MIME type.
    @discardableResult
    public func addSource(_ url: String, _ mimeType: String) -> MediaBuilder {
        var sources = data["sources"] as! [Any]
        sources.append(["url": url, "mimeType": mimeType] as [String: Any])
        data["sources"] = sources
        return self
    }

    /// Adds a pre-built source map directly.
    @discardableResult
    public func addSourceMap(_ source: [String: Any]) -> MediaBuilder {
        var sources = data["sources"] as! [Any]
        sources.append(source)
        data["sources"] = sources
        return self
    }

    public func build() -> Card {
        return data
    }
}

File: rust/fluentcards/src/builders/media_builder.rs
use serde_json::Value;

use crate::enums::Spacing;
use crate::models::Card;

/// Builds a Media Adaptive Card element.
pub struct MediaBuilder {
    data: Card,
}

impl MediaBuilder {
    pub(crate) fn new() -> Self {
        let mut data = Card::new();
        data.insert("type".into(), Value::String("Media".into()));
        data.insert("sources".into(), Value::Array(Vec::new()));
        MediaBuilder { data }
    }

    pub fn with_id(&mut self, id: &str) -> &mut Self {
        self.data.insert("id".into(), Value::String(id.into()));
        self
    }

    pub fn with_poster(&mut self, poster: &str) -> &mut Self {
        self.data
            .insert("poster".into(), Value::String(poster.into()));
        self
    }

    pub fn with_alt_text(&mut self, alt_text: &str) -> &mut Self {
        self.data
            .insert("altText".into(), Value::String(alt_text.into()));
        self
    }

    pub fn with_spacing(&mut self, spacing: Spacing) -> &mut Self {
        self.data.insert("spacing".into(), spacing.into());
        self
    }

    /// Adds a media source with the given URL and MIME type.
    pub fn add_source(&mut self, url: &str, mime_type: &str) -> &mut Self {
        let mut source = Card::new();
        source.insert("url".into(), Value::String(url.into()));
        source.insert("mimeType".into(), Value::String(mime_type.into()));
        if let Some(Value::Array(sources)) = self.data.get_mut("sources") {
            sources.push(Value::Object(source));
        }
        self
    }

    /// Adds a pre-built source map directly.
    pub fn add_source_map(&mut self, source: Card) -> &mut Self {
        if let Some(Value::Array(sources)) = self.data.get_mut("sources") {
            sources.push(Value::Object(source));
        }
        self
    }

    pub fn build(&self) -> Card {
        self.data.clone()
    }
}
